From ad60f3d14891097b0cb8ba718e7e6ccdbd43654b Mon Sep 17 00:00:00 2001
From: Nirbheek Chauhan <nirbheek@centricular.com>
Date: Wed, 3 Jun 2020 04:45:42 +0530
Subject: [PATCH 11/24] gwin32: Stub out all version checking on WinRT

Also remove all `WINAPI_FAMILY == 2` conditionals. That code is
equivalent to `G_WINAPI_FAMILY_APP` being defined, and we now no
longer run any version checking in that case.
---
 glib/gwin32.c | 70 ++++++++++++++++++++++++++++++++++-----------------
 1 file changed, 47 insertions(+), 23 deletions(-)

diff --git a/glib/gwin32.c b/glib/gwin32.c
index ab9e4b571..1bd754fe7 100644
--- a/glib/gwin32.c
+++ b/glib/gwin32.c
@@ -40,35 +40,19 @@
 #include <wchar.h>
 #include <errno.h>
 #include <fcntl.h>
+#include <ctype.h>
 
 #define STRICT			/* Strict typing, please */
 #include <winsock2.h>
 #include <windows.h>
 #undef STRICT
+
 #ifndef G_WITH_CYGWIN
 #include <direct.h>
 #endif
-#include <errno.h>
-#include <ctype.h>
-#if defined(_MSC_VER) || defined(__DMC__)
-#  include <io.h>
-#endif /* _MSC_VER || __DMC__ */
-
-#define MODERN_API_FAMILY 2
-
-#if WINAPI_FAMILY == MODERN_API_FAMILY
-/* This is for modern UI Builds, where we can't use LoadLibraryW()/GetProcAddress() */
-/* ntddk.h is found in the WDK, and MinGW */
-#include <ntddk.h>
 
 #ifdef _MSC_VER
-#pragma comment (lib, "ntoskrnl.lib")
-#endif
-#elif defined(__MINGW32__) && !defined(__MINGW64_VERSION_MAJOR)
-/* mingw-w64 must use winternl.h, but not MinGW */
-#include <ntdef.h>
-#else
-#include <winternl.h>
+#include <io.h>
 #endif
 
 #include "glib.h"
@@ -132,6 +116,11 @@ g_win32_getlocale (void)
       || ((ev = g_getenv ("LANG")) != NULL && ev[0] != '\0'))
     return g_strdup (ev);
 
+#ifdef G_WINAPI_ONLY_APP
+  g_critical ("g_win32_getlocale() has not been fully implemented: "
+              "GLib was built for Universal Windows Platform apps");
+  return g_strdup ("C");
+#else
   lcid = GetThreadLocale ();
 
   if (!GetLocaleInfoW (lcid, LOCALE_SISO639LANGNAME, iso639, sizeof (iso639)) ||
@@ -181,6 +170,7 @@ g_win32_getlocale (void)
       break;
     }
 
+#endif
   iso639_utf8 = g_utf16_to_utf8 (iso639, -1, NULL, NULL, NULL);
   iso3166_utf8 = g_utf16_to_utf8 (iso3166, -1, NULL, NULL, NULL);
 
@@ -325,6 +315,9 @@ g_win32_get_package_installation_directory_of_module (gpointer hmodule)
   return retval;
 }
 
+/* Package directories do not make sense on UWP, and the required API is also
+ * not available */
+#ifndef G_WINAPI_ONLY_APP
 static gchar *
 get_package_directory_from_module (const gchar *module_name)
 {
@@ -373,6 +366,7 @@ get_package_directory_from_module (const gchar *module_name)
 
   return g_strdup (fn);
 }
+#endif /* !G_WINAPI_ONLY_APP */
 
 /**
  * g_win32_get_package_installation_directory:
@@ -439,11 +433,17 @@ g_win32_get_package_installation_directory (const gchar *package,
   if (package != NULL)
       g_warning ("Passing a non-NULL package to g_win32_get_package_installation_directory() is deprecated and it is ignored.");
 
+#ifdef G_WINAPI_ONLY_APP
+  g_warning ("g_win32_get_package_installation_directory() will not work since "
+             "GLib was built for Universal Windows Platform apps");
+#else
+
   if (dll_name != NULL)
     result = get_package_directory_from_module (dll_name);
 
   if (result == NULL)
     result = get_package_directory_from_module (NULL);
+#endif
 
   return result;
 }
@@ -499,6 +499,7 @@ G_GNUC_END_IGNORE_DEPRECATIONS
  * private API to call Windows's RtlGetVersion(), which may need to be called
  * via GetProcAddress()
  */
+#ifndef G_WINAPI_ONLY_APP
 gboolean
 _g_win32_call_rtl_version (OSVERSIONINFOEXW *info)
 {
@@ -509,7 +510,6 @@ _g_win32_call_rtl_version (OSVERSIONINFOEXW *info)
 
   if (g_once_init_enter (&inited))
     {
-#if WINAPI_FAMILY != MODERN_API_FAMILY
       /* For non-modern UI Apps, use the LoadLibraryW()/GetProcAddress() thing */
       typedef NTSTATUS (WINAPI fRtlGetVersion) (PRTL_OSVERSIONINFOEXW);
 
@@ -519,16 +519,14 @@ _g_win32_call_rtl_version (OSVERSIONINFOEXW *info)
 
       RtlGetVersion = (fRtlGetVersion *) GetProcAddress (hmodule, "RtlGetVersion");
       g_return_val_if_fail (RtlGetVersion != NULL, FALSE);
-#endif
 
       memset (&result, 0, sizeof (OSVERSIONINFOEXW));
       result.dwOSVersionInfoSize = sizeof (OSVERSIONINFOEXW);
 
       RtlGetVersion (&result);
 
-#if WINAPI_FAMILY != MODERN_API_FAMILY
       FreeLibrary (hmodule);
-#endif
+
       g_once_init_leave (&inited, TRUE);
     }
 
@@ -536,6 +534,7 @@ _g_win32_call_rtl_version (OSVERSIONINFOEXW *info)
 
   return TRUE;
 }
+#endif /* !G_WINAPI_ONLY_APP */
 
 /**
  * g_win32_check_windows_version:
@@ -568,6 +567,13 @@ g_win32_check_windows_version (const gint major,
                                const gint spver,
                                const GWin32OSType os_type)
 {
+#ifdef G_WINAPI_ONLY_APP
+  /* RtlGetVersion is not available when building for UWP, but we only support
+   * Windows 10 or newer in that case, so we can hard-code the results */
+  if (os_type != G_WIN32_OS_SERVER && major >= 10)
+    return TRUE;
+  return FALSE;
+#else
   OSVERSIONINFOEXW osverinfo;
   gboolean is_ver_checked = FALSE;
   gboolean is_type_checked = FALSE;
@@ -619,6 +625,7 @@ g_win32_check_windows_version (const gint major,
     }
 
   return is_ver_checked && is_type_checked;
+#endif
 }
 
 /**
@@ -650,8 +657,14 @@ g_win32_get_windows_version (void)
 {
   static gsize windows_version;
 
+#ifdef G_WINAPI_ONLY_APP
+  /* GetVersion() is not available when building for UWP, so just always return
+   * 62 as documented above */
+  windows_version = 0x00000206;
+#else
   if (g_once_init_enter (&windows_version))
     g_once_init_leave (&windows_version, GetVersion ());
+#endif
 
   return windows_version;
 }
@@ -741,6 +754,7 @@ g_win32_locale_filename_from_utf8 (const gchar *utf8filename)
 
   retval = special_wchar_to_locale_encoding (wname);
 
+#ifndef G_WINAPI_ONLY_APP
   if (retval == NULL)
     {
       /* Conversion failed, so check if there is a 8.3 version, and use that. */
@@ -749,6 +763,7 @@ g_win32_locale_filename_from_utf8 (const gchar *utf8filename)
       if (GetShortPathNameW (wname, wshortname, G_N_ELEMENTS (wshortname)))
         retval = special_wchar_to_locale_encoding (wshortname);
     }
+#endif
 
   g_free (wname);
 
@@ -796,6 +811,10 @@ g_win32_locale_filename_from_utf8 (const gchar *utf8filename)
 gchar **
 g_win32_get_command_line (void)
 {
+#ifdef G_WINAPI_ONLY_APP
+  g_error ("g_win32_get_command_line() is not available since "
+           "GLib was built for Universal Windows Platform apps");
+#else
   gchar **result;
   LPWSTR *args;
   gint i, n;
@@ -809,6 +828,7 @@ g_win32_get_command_line (void)
 
   LocalFree (args);
   return result;
+#endif
 }
 
 /* Binary compatibility versions. Not for newly compiled code. */
@@ -1324,14 +1344,18 @@ g_crash_handler_win32_init (void)
   else
     debugger_spawn_flags = CREATE_NEW_CONSOLE;
 
+#ifndef G_WINAPI_ONLY_APP
   WinVEH_handle = AddVectoredExceptionHandler (0, &g_win32_veh_handler);
+#endif
 }
 
 void
 g_crash_handler_win32_deinit (void)
 {
+#ifndef G_WINAPI_ONLY_APP
   if (WinVEH_handle != NULL)
     RemoveVectoredExceptionHandler (WinVEH_handle);
+#endif
 
   WinVEH_handle = NULL;
 }
-- 
2.44.0.windows.1

